### DISCUSSION TITLE
How to use prebuilt, dynamic library on Windows

### DISCUSSION THREAD

**Discussion by tastewar** [ned14/llfio#147] [Q&A]
When building a project that incorporates llfio on Windows, I have the project setup to link against the import libraries that were included, llfio and ntkernel-error-category, but when I attempt to build, during the link phase I get 2 unresolved externals. The names are well decorated, so I won't paste them here, but one appears to be from quickcpplib and the other from outcome. I don't see either import libraries or dlls for these in the llfio pre-built download, and I don't see them available in their root projects either.

**Comment by ned14** [comment]
quickcpplib and outcome are header only libraries.

It would be useful if you did post the unresolved externals. Usually it is due to using a different C++ standard in the final binary than was used to build the libraries.

**Reply by tastewar** [reply]
[Build:x64d] raflibii.lib(raflib.obj) : error LNK2019: unresolved external symbol "__declspec(dllimport) class quickcpplib::_808dfbc9::ringbuffer_log::ringbuffer_log<struct quickcpplib::_808dfbc9::ringbuffer_log::simple_ringbuffer_log_policy<1048576>,struct llfio_v2_e7b0ca82::detail::thread_local_log_level_filter> & __cdecl llfio_v2_e7b0ca82::log(void)" (__imp_?log@llfio_v2_e7b0ca82@@YAAEAV?$ringbuffer_log@U?$simple_ringbuffer_log_policy@$0BAAAAA@@0_808dfbc9@quickcpplib@@Uthread_local_log_level_filter@detail@llfio_v2_e7b0ca82@@@ringbuffer_log@_808dfbc9@quickcpplib@@XZ) referenced in function "void __cdecl llfio_v2_e7b0ca82::detail::append_path_info<struct llfio_v2_e7b0ca82::error_info const >(struct llfio_v2_e7b0ca82::error_info const &,class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > &)" (??$append_path_info@$$CBUerror_info@llfio_v2_e7b0ca82@@@detail@llfio_v2_e7b0ca82@@YAXAEBUerror_info@1@AEAV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@@Z) [C:\projects\raflibii\Debug\rltest.vcxproj]
[Build:x64d] raflibii.lib(raflib.obj) : error LNK2019: unresolved external symbol "__declspec(dllimport) public: static class outcome_v2::basic_result<class llfio_v2_e7b0ca82::file_handle,struct llfio_v2_e7b0ca82::error_info,struct outcome_v2::policy::error_code_throw_as_system_error<class llfio_v2_e7b0ca82::file_handle,struct llfio_v2_e7b0ca82::error_info,void> > __cdecl llfio_v2_e7b0ca82::file_handle::file(class llfio_v2_e7b0ca82::path_handle const &,class llfio_v2_e7b0ca82::path_view,enum llfio_v2_e7b0ca82::handle::mode,enum llfio_v2_e7b0ca82::handle::creation,enum llfio_v2_e7b0ca82::handle::caching,struct quickcpplib::_808dfbc9::bitfield::bitfield<struct llfio_v2_e7b0ca82::handle::flag_base>)" (__imp_?file@file_handle@llfio_v2_e7b0ca82@@SA?AV?$basic_result@Vfile_handle@llfio_v2_e7b0ca82@@Uerror_info@2@U?$error_code_throw_as_system_error@Vfile_handle@llfio_v2_e7b0ca82@@Uerror_info@2@X@policy@outcome_v2@@@outcome_v2@@AEBVpath_handle@2@Vpath_view@2@W4mode@handle@2@W4creation@82@W4caching@82@U?$bitfield@Uflag_base@handle@llfio_v2_e7b0ca82@@@bitfield@_808dfbc9@quickcpplib@@@Z) referenced in function "class outcome_v2::basic_result<class llfio_v2_e7b0ca82::file_handle,struct llfio_v2_e7b0ca82::error_info,struct outcome_v2::policy::error_code_throw_as_system_error<class llfio_v2_e7b0ca82::file_handle,struct llfio_v2_e7b0ca82::error_info,void> > __cdecl llfio_v2_e7b0ca82::file(class llfio_v2_e7b0ca82::path_handle const &,class llfio_v2_e7b0ca82::path_view,enum llfio_v2_e7b0ca82::handle::mode,enum llfio_v2_e7b0ca82::handle::creation,enum llfio_v2_e7b0ca82::handle::caching,struct quickcpplib::_808dfbc9::bitfield::bitfield<struct llfio_v2_e7b0ca82::handle::flag_base>)" (?file@llfio_v2_e7b0ca82@@YA?AV?$basic_result@Vfile_handle@llfio_v2_e7b0ca82@@Uerror_info@2@U?$error_code_throw_as_system_error@Vfile_handle@llfio_v2_e7b0ca82@@Uerror_info@2@X@policy@outcome_v2@@@outcome_v2@@AEBVpath_handle@1@Vpath_view@1@W4mode@handle@1@W4creation@71@W4caching@71@U?$bitfield@Uflag_base@handle@llfio_v2_e7b0ca82@@@bitfield@_808dfbc9@quickcpplib@@@Z) [C:\projects\raflibii\Debug\rltest.vcxproj]
[Build:x64d] C:\projects\raflibii\Debug\bin\rltest.exe : fatal error LNK1120: 2 unresolved externals [C:\projects\raflibii\Debug\rltest.vcxproj]

**Reply by tastewar** [reply]
There doesn't seem to be an import library for quickcpplib or outcome in the pre-built download. (or the corresponding dll)

**Reply by tastewar** [reply]
I understand that those are supposed to be header only, but something in there has caused them to be built in such a way that the linker is looking for functions that are __declspec(dllimport)

**Reply by ned14** [reply]
Are you compiling your program using `target_link_libraries()` to the llfio cmake target?

**Reply by tastewar** [reply]
I am using the headers downloaded from https://github.com/ned14/llfio/archive/refs/tags/20240823.zip and linking in the import libs downloaded from https://github.com/ned14/llfio/releases/download/20240823/llfio-v2.0-binaries-win64-vs2022-error_code.zip. In my project, I have configured my include directories to point to the downloaded include folders. In my source, I am defining the following:

#define NTKERNEL_ERROR_CATEGORY_INLINE 0
#define LLFIO_EXCLUDE_DYNAMIC_THREAD_POOL_GROUP
#define LLFIO_EXCLUDE_NETWORKING

before including llfio.hpp.

**Reply by ned14** [reply]
Something odd is going on there. The prebuilt binaries distro is prepared after a complete build and link. The binaries worked at at that point.

Assuming you don't want to dig into exported symbols and matching up what is in the shared library export list and the unresolved symbols, I'd suggest reverting to building this locally.

You can add llfio as a git submodule, add it as a subdirectory from within cmake, and link to its exported targets. Should all "just work".

**Reply by tastewar** [reply]
I am following the build instructions for Windows on https://ned14.github.io/llfio/md__build.html . I did the git config --system core.longpaths true and it gave me an error, so I did it from an admin shell, and then it worked. Now in a normal powershell, if I execute git config --list it shows core.longpaths=true. I then did the recursive clone, created the build directory, switched to it, and ran cmake .. without specifying a generator, and it echoed

-- Building for: Visual Studio 17 2022

But at the beginning of building outcome it reported this error:

      1>Checking Build System

  C:\Program Files\Microsoft Visual
  Studio\2022\Professional\MSBuild\Microsoft\VC\v170\Microsoft.CppBuild.targets(382,5):
  error MSB3491: Could not write lines to file
  "outcome-link-test-experimental-c-result-type_SHARED-vis_hidden-cxx-helpers.dir\Debug\outcome-.D6F5F0D7.tlog\outcome-link-test-experimental-c-result-type_SHARED-vis_hidden-cxx-helpers.lastbuildstate".
  Path:
  outcome-link-test-experimental-c-result-type_SHARED-vis_hidden-cxx-helpers.dir\Debug\outcome-.D6F5F0D7.tlog\outcome-link-test-experimental-c-result-type_SHARED-vis_hidden-cxx-helpers.lastbuildstate
  exceeds the OS max path limit.  The fully qualified file name must be less
  than 260 characters.
  [C:\projects\llfio\build\outcome\repo\src\outcome-build\test\link\outcome-link-test-experimental-c-result-type_SHARED-vis_hidden-cxx-helpers.vcxproj]
  [C:\projects\llfio\build\outcome\outcome.vcxproj]

**Reply by tastewar** [reply]
I got past this by enabling long file names in the registry. I can't understand why this isn't the default on Windows 11. Ugh. Anyway, on to the next error, which I don't think is related to my environment. These are unresolved external errors when linking (some binary in outcome, I believe)

         Creating library C:/projects/llfio/build/outcome/repo/src/outcome-build/test/link/Debug/outcome-link-test-experimental-c-result-type_SHARED-vis_default-lib.lib and object C:/projects/llfio/build/outcome/repo/src/outcome-build/test/link/Debug/outcome-link-test-experimental-c-result-type_SHARED-vis_default-lib.exp
      outcome-link-test-experimental-c-result-type_SHARED-vis_default-lib.vcxproj -> C:\projects\llfio\build\outcome\repo\src\outcome-build\bin\Debug\outcome-link-test-experimental-c-result-type_SHARED-vis_default-lib.dll
      Building Custom Rule C:/projects/llfio/build/outcome/repo/src/outcome/test/link/CMakeLists.txt
      main.c

  main.obj : error LNK2019: unresolved external symbol
  outcome_make_result_status_code_failure_posix referenced in function
  outcome_make_result_system_mylib_failure_posix
  [C:\projects\llfio\build\outcome\repo\src\outcome-build\test\link\outcome-link-test-experimental-c-result-type_SHARED-vis_default.vcxproj]
  [C:\projects\llfio\build\outcome\outcome.vcxproj]

  main.obj : error LNK2019: unresolved external symbol
  outcome_make_result_status_code_failure_system referenced in function
  outcome_make_result_system_mylib_failure_system
  [C:\projects\llfio\build\outcome\repo\src\outcome-build\test\link\outcome-link-test-experimental-c-result-type_SHARED-vis_default.vcxproj]
  [C:\projects\llfio\build\outcome\outcome.vcxproj]

  main.obj : error LNK2019: unresolved external symbol
  outcome_status_code_message referenced in function result_failure_message
  [C:\projects\llfio\build\outcome\repo\src\outcome-build\test\link\outcome-link-test-experimental-c-result-type_SHARED-vis_default.vcxproj]
  [C:\projects\llfio\build\outcome\outcome.vcxproj]


  C:\projects\llfio\build\outcome\repo\src\outcome-build\bin\Debug\outcome-link-test-experimental-c-result-type_SHARED-vis_default.exe
  : fatal error LNK1120: 3 unresolved externals
  [C:\projects\llfio\build\outcome\repo\src\outcome-build\test\link\outcome-link-test-experimental-c-result-type_SHARED-vis_default.vcxproj]
  [C:\projects\llfio\build\outcome\outcome.vcxproj]

**Reply by ned14** [reply]
Sorry it's been so much of a chore for you.

Your last errors look like those currently failing LLFIO on the Windows CI. I will look into them later today as usefully I am currently at a WG21 standards meeting, so I have the time.

**Reply by tastewar** [reply]
Thank you, Niall. And thank you for the fix on the Linux side. Ironically, we only discovered that due to having to work on the Linux side due to my challenges in getting LLFIO integrated on the Windows side ;-)

But mostly, I thank you for providing this library. It is exactly what we needed for our project. We are re-writing a database engine that has been Windows based for the past 30 years to run on Linux. We have 30 years of experience writing code with unbuffered i/o on Windows, but have been trying to re-engineer our software to run on both Windows and Linux. I had researched how to do the same things on Linux, and it's not that it's difficult, it's just that I would rather that our source code was cross-platform, and the platform specific implementation was done by someone else (competent). So thank you for being that person. Writing the code to deal with sparse files was so much easier with LLFIO.

I am attempting to build from the develop branch, partly because I knew we'd be wanting the Linux sparse file fix. If you are saying that this build problem is known, and that I could switch to the 20240823 tag and expect success, I will start with that.

**Reply by tastewar** [reply]
On Linux, 20240823 builds cleanly for me, but the develop branch does not. I get 2 errors reported like this:

In file included from /home/tastewar/xp/llfio/src/../include/llfio/v2.0/file_handle.hpp:562,
                 from /home/tastewar/xp/llfio/src/../include/llfio/v2.0/fast_random_file_handle.hpp:28,
                 from /home/tastewar/xp/llfio/src/../include/llfio/v2.0/llfio.hpp:69,
                 from /home/tastewar/xp/llfio/src/../include/llfio/llfio.hpp:18,
                 from /home/tastewar/xp/llfio/src/llfio.cpp:7:
/home/tastewar/xp/llfio/src/../include/llfio/v2.0/detail/impl/posix/file_handle.ipp: In member function ‘virtual llfio_v2_27d96a8a::result<long long unsigned int> llfio_v2_27d96a8a::file_handle::zero(llfio_v2_27d96a8a::file_handle::extent_pair, llfio_v2_27d96a8a::deadline)’:
/home/tastewar/xp/llfio/src/../include/llfio/v2.0/detail/impl/posix/file_handle.ipp:997:12: error: could not convert ‘extent’ from ‘llfio_v2_27d96a8a::file_handle::extent_pair’ to ‘llfio_v2_27d96a8a::result<long long unsigned int>’ {aka ‘outcome_v2::basic_result<long long unsigned int, llfio_v2_27d96a8a::error_info, outcome_v2::policy::error_code_throw_as_system_error<long long unsigned int, llfio_v2_27d96a8a::error_info, void> >’}
  997 |     return extent;
      |            ^~~~~~
      |            |
      |            llfio_v2_27d96a8a::file_handle::extent_pair

**Reply by tastewar** [reply]
On Windows, attempting to build 20240823 also fails. The "cmake .." step lists a FATAL error early, but carries on. Here is that error:

-- Missing dependency quickcpplib is NOT installed in cmake packages!
-- Missing dependency quickcpplib is NOT found at C:/projects/llfio/build/quickcpplib!
-- Superbuilding missing dependency quickcpplib with config Debug, this may take a while ...
-- Missing dependency outcome is NOT installed in cmake packages!
-- Missing dependency outcome is NOT found at C:/projects/llfio/build/outcome!
-- Superbuilding missing dependency outcome with config Debug, this may take a while ...
CMake Error at build/quickcpplib/repo/cmakelib/QuickCppLibUtils.cmake:95 (message):
  FATAL: Build download, build and install of outcome with
  -DCMAKE_BUILD_TYPE=Debug -G "Visual Studio 17 2022" -DBUILD_TESTING=OFF
  "-DQUICKCPPLIB_ROOT_BINARY_DIR=C:/projects/llfio/build" failed with error
  '1'

But then later on (after the FATAL error...), I get 3 unresolved externals from main.obj when building outcome. I believe this is the same as in the develop branch:

  main.obj : error LNK2019: unresolved external symbol
  outcome_make_result_status_code_failure_posix referenced in function
  outcome_make_result_system_mylib_failure_posix
  [C:\projects\llfio\build\outcome\repo\src\outcome-build\test\link\outcome-link-test-experimental-c-result-type_SHARED-vis_default.vcxproj]
  [C:\projects\llfio\build\outcome\outcome.vcxproj]

  main.obj : error LNK2019: unresolved external symbol
  outcome_make_result_status_code_failure_system referenced in function
  outcome_make_result_system_mylib_failure_system
  [C:\projects\llfio\build\outcome\repo\src\outcome-build\test\link\outcome-link-test-experimental-c-result-type_SHARED-vis_default.vcxproj]
  [C:\projects\llfio\build\outcome\outcome.vcxproj]

  main.obj : error LNK2019: unresolved external symbol
  outcome_status_code_message referenced in function result_failure_message
  [C:\projects\llfio\build\outcome\repo\src\outcome-build\test\link\outcome-link-test-experimental-c-result-type_SHARED-vis_default.vcxproj]
  [C:\projects\llfio\build\outcome\outcome.vcxproj]


  C:\projects\llfio\build\outcome\repo\src\outcome-build\bin\Debug\outcome-link-test-experimental-c-result-type_SHARED-vis_default.exe
  : fatal error LNK1120: 3 unresolved externals
  [C:\projects\llfio\build\outcome\repo\src\outcome-build\test\link\outcome-link-test-experimental-c-result-type_SHARED-vis_default.vcxproj]
  [C:\projects\llfio\build\outcome\outcome.vcxproj]

**Reply by ned14** [reply]
Both the Linux build and Windows builds should now be fixed. Thanks for your patience.

I'm not allowed merge to master branch until the Boost release happens next month. Until then, anything fetching master branch under the bonnet will continue to fail.

Once Boost releases, I generally do a github release of everything here.

**Reply by tastewar** [reply]
Thank you. I can build both Windows and Linux now locally. Still struggling with linking everything into a single Windows exe, but still plugging away at it. Linux links just fine.
